src/zkevm_specs/evm/execution/codecopy.py

```diff
@@ -0,0 +1,53 @@
+from ...util import N_BYTES_MEMORY_ADDRESS, FQ
+from ..execution_state import ExecutionState
+from ..instruction import Instruction, Transition
+from ..step import CopyCodeToMemoryAuxData
+from ..table import RW, RWTableTag, CallContextFieldTag, AccountFieldTag
+
+
+def codecopy(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    memory_offset_word, code_offset_word, size_word = (
+        instruction.stack_pop(),
+        instruction.stack_pop(),
+        instruction.stack_pop(),
+    )
+
+    memory_offset, size = instruction.memory_offset_and_length(memory_offset_word, size_word)
+    code_offset = instruction.rlc_to_fq_exact(code_offset_word, N_BYTES_MEMORY_ADDRESS)
+
+    account = instruction.call_context_lookup(CallContextFieldTag.CalleeAddress)
+    code_size = instruction.account_read(account, AccountFieldTag.CodeSize)
+    code_hash = instruction.account_read(account, AccountFieldTag.CodeHash)
+
+    next_memory_size, memory_expansion_gas_cost = instruction.memory_expansion_dynamic_length(
+        memory_offset, size
+    )
+    gas_cost = instruction.memory_copier_gas_cost(size, memory_expansion_gas_cost)
+
+    if not instruction.is_zero(size):
```

Comment:
I'm missing a call to `CopyCodeToMemory` or `BufferedReader` gadget here. Otherwise, you're just constraining the offsets and lengths, but not actually copying anything no?

Maybe I'm missing something.

Reply:
@CPerezz We check that if the "to be copied" `size` is non-zero, then the next execution state MUST BE the internal state `CopyCodeToMemory`. We add a few more constraints on the auxiliary data, but the rest is taken over by the `CopyCodeToMemory` gadget.